### ISSUE TITLE
Tensor.max()/min() over multiple axes

### ISSUE THREAD

**Issue by f0k** [pytorch/pytorch#28213]
## 🚀 Feature

`Tensor.max()` and `Tensor.min()` support a `dim` argument to reduce over a single axis. They should support multiple axes just like `Tensor.sum()`.

## Motivation

Sometimes I want to pool over all spatial dimensions, or an arbitrary set of dimensions. Currently I have to decide myself whether to loop over dimensions (and in which order), or reshape the tensor to collapse multiple dimensions into one.

## Pitch

`max()` and `min()` should support a list of dimensions. In this case, the returned indices tensor would be a matrix giving the coordinates along each dimension in their given order.

## Alternatives

For the specific case of max-pooling across spatial dimensions, I could use torch.nn.functional.adaptive_max_pool2d with an output size of 1.

When extending `max()` and `min()`, there are different options for the returned indices tensor, ordered from easiest to most complex implementation:
1. not returning any indices if there are multiple dimensions
2. returning a vector of indices that index into a flattened view of the dimensions to reduce (this is what adaptive_max_pool2d_with_indices does)
3. returning a matrix of indices that give the indices along each of the reduced dimensions, in ascending order of dimensions
4. returning a matrix of indices that give the indices along each of the reduced dimensions, in the given order of dimensions

The first would avoid some headaches, but cannot be undone again without breaking the signature. It would also complicate user code that has to work both with single and multiple dimensions. So that's probably not a good route.
The second would be compatible to adaptive max-pooling, but other than that, not very practical to deal with (think of advanced indexing).
The third and forth are similarly useful. Most of the time they will be equivalent. I don't really know what users would expect who provide `dim` in a not-ascending order.

We cannot really look at numpy for guidance -- `np.max()` supports multiple axes, but `np.argmax()` doesn't. If `np.argmax()` is called without specifying an axis, it will reduce over all axes and return an index into the flattened array (i.e., option 2, the one I deemed impractical).

Disclaimer: I don't need the indices anyway.

## Additional context

For summation, this was suggested in #2006 and implemented in #6152.
There also is a generic multi-dimensional operators issue in #9703, but the indices warrant some extra discussion.

**Comment by smessmer** [comment]
This sounds like a good idea to me. cc @gchanan for guidance on what the returned indices should be.

**Comment by harshsha5** [comment]
This feature would be super useful!

**Comment by zasdfgbnm** [comment]
See also https://github.com/pytorch/pytorch/issues/31042. This feature is being implemented at https://github.com/pytorch/pytorch/pull/43092.

**Comment by f0k** [comment]
Thanks @zasdfgbnm for implementing it! :tada: A second set of functions that does not return indices was a good way out. Closing this as #43092 is already merged.
